projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-top-posts.php

```diff
@@ -0,0 +1,201 @@
+		// This is to minimise requests when it can already be handled by the block.
+		register_rest_route(
+			'wpcom/v2',
+			'/top-posts',
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_top_posts' ),
+					'permission_callback' => '__return_true',
+					'args'                => array(
+						'period' => array(
+							'description'       => __( 'Timeframe for stats.', 'jetpack' ),
+							'type'              => array( 'string', 'integer' ),
+							'required'          => true,
+							'validate_callback' => function ( $param ) {
+								return is_numeric( $param ) || is_string( $param );
+							},
+						),
+						'number' => array(
+							'description'       => __( 'Number of posts to display.', 'jetpack' ),
+							'type'              => 'integer',
+							'required'          => false,
+							'validate_callback' => function ( $param ) {
+								return is_numeric( $param );
+							},
+						),
+						'types'  => array(
+							'description'       => __( 'Types of content to include.', 'jetpack' ),
+							'type'              => 'string',
+							'required'          => false,
+							'validate_callback' => function ( $param ) {
+								return is_string( $param );
+							},
+						),
+					),
+				),
+			)
+		);
+	}
+
+	/**
+	 * Get the site's post types.
+	 *
+	 * @return array Site's post types.
+	 */
+	public function get_post_types() {
+		$post_types       = array_values( get_post_types( array( 'public' => true ) ) );
+		$post_types_array = array();
+
+		foreach ( $post_types as $type ) {
+			$post_types_array[] = array(
+				'label' => get_post_type_object( $type )->labels->name,
+				'id'    => $type,
+			);
+		}
+
+		return $post_types_array;
+	}
+
+	/**
+	 * Get the site's top content.
+	 *
+	 * @param \WP_REST_Request $request request object.
+	 *
+	 * @return array Data on top posts.
+	 */
+	public function get_top_posts( $request ) {
+		$period        = $request->get_param( 'period' );
+		$all_time_days = floor( ( time() - strtotime( get_option( 'site_created_date' ) ) ) / ( 60 * 60 * 24 * 365 ) );
+
+		// While we only display ten posts, users can filter out content types.
+		// As such, we should obtain a few spare posts from the Stats endpoint.
+		$posts_to_obtain_count = 30;
+
+		// We should not override cache when displaying the block on the frontend.
+		// But we should allow instant preview of changes when editing the block.
+		$is_rendering_block = isset( $request['types'] );
+		$override_cache     = ! $is_rendering_block;
+
+		$query_args = array(
+			'max'       => $posts_to_obtain_count,
+			'summarize' => true,
+			'num'       => $period !== 'all-time' ? $period : $all_time_days,
+			'period'    => 'day',
+		);
+
+		$data            = ( new WPCOM_Stats() )->get_top_posts( $query_args, $override_cache );
+		$posts_retrieved = count( $data['summary']['postviews'] );
```

Comment:
It's possible that the $data is an instance of WP_Error, so it'll be great if we could give it a check and resolve the flow here, for example:

```suggestion
		if ( is_wp_error( $data ) ) {
			$data['summary']['postviews'] = array();
		}
		$posts_retrieved = count( $data['summary']['postviews'] );
```

Reply:
Added!

Reply:
Hi @Aurorum Thanks for adressing it. My bad, `$data `is still an object there and probably we should do this:

```PHP
$data = array( 'summary' => array( 'postviews' => array() ) );
```

Reply:
Whoops, fixed!